File: 2024/src/day06/day06.py
from functools import cache


def processing() -> dict[complex, str]:
    return {i + j * 1j: c for i, r in enumerate(open("src/day06/input.txt").readlines()) for j, c in
            enumerate(r.strip())}


def print_grid(grid: dict[complex, str]) -> None:
    for i in range(10):
        for j in range(10):
            print(grid[(i + j * 1j)], end="")
        print()


def out(grid: dict[complex, str], start: complex) -> set[complex]:
    p, d, visited = start, -1, set()
    while p in grid and (p, d) not in visited:
        visited.add((p, d))
        if grid.get(p + d) == "#":
            d *= -1j
        else:
            p += d
    return {p for p, _ in visited}


def part1() -> int:
    grid: dict[complex, str] = processing()
    return len(out(grid, [p for p in grid if grid[p] == '^'][0]))


@cache
def part2() -> int:
    start_grid: dict[complex, str] = processing()
    start: complex = [p for p in start_grid if start_grid[p] == '^'][0]
    count = 0
    for o in out(start_grid, start):
        if o == start:
            continue
        grid = start_grid | {o: "#"}
        p, d, visited = start, -1, set()
        while p in grid and (p, d) not in visited:
            visited.add((p, d))
            if grid.get(p + d) == "#":
                d *= -1j
            else:
                p += d
        count += (p, d) in visited
    return count


if __name__ == "__main__":
    print(part1())
    print(part2())

File: 2023/src/day06/day06.go
package main

import (
	"2023/src/utils"
	"fmt"
	"math"
	"strconv"
	"strings"
)

func processing(filename string) ([][]int, []int) {
	tab1 := make([][]int, 2)
	tab2 := make([]int, 2)
	timeDistance := strings.Split(utils.ReadFile(filename), "\n")
	timeText := strings.Replace(strings.Replace(strings.Replace(strings.Trim(timeDistance[0][9:], " "), "  ", " ", -1), "  ", " ", -1), "  ", " ", -1)
	distanceText := strings.Replace(strings.Replace(strings.Replace(strings.Trim(timeDistance[1][9:], " "), "  ", " ", -1), "  ", " ", -1), "  ", " ", -1)
	tab1[0] = utils.Map2(strings.SplitN(timeText, " ", -1), func(str string) int {
		number, err := strconv.Atoi(str)
		if err != nil {
			panic(err)
		}
		return number
	})
	tab1[1] = utils.Map2(strings.SplitN(distanceText, " ", -1), func(str string) int {
		number, err := strconv.Atoi(str)
		if err != nil {
			panic(err)
		}
		return number
	})
	tab2[0], _ = strconv.Atoi(strings.Replace(timeText, " ", "", -1))
	tab2[1], _ = strconv.Atoi(strings.Replace(distanceText, " ", "", -1))
	return tab1, tab2
}

func part1(filename string) int {
	result := 1
	tab, _ := processing(filename)
	races := utils.Zip(tab[0], tab[1])
	for _, race := range races {
		count := 0
		for hold := 0; hold < race.First; hold++ {
			if hold*(race.First-hold) > race.Second {
				count++
			}
		}
		result *= count
	}
	return result
}

func part2(filename string) int {
	_, tab := processing(filename)
	delta := tab[0]*tab[0] - 4*tab[1]
	if delta == 0 {
		return 1
	} else if delta > 0 {
		x1 := int((-float64(tab[0]) + math.Sqrt(float64(delta))) / 2)
		x2 := int((-float64(tab[0]) - math.Sqrt(float64(delta))) / 2)
		return x1 - x2
	}
	return 0
}

func main() {
	filename := "day06/input.txt"
	fmt.Println(part1(filename))
	fmt.Println(part2(filename))
}
